File: quickstart_guides/arrays_strings/java/IsPalindrome.java
/*
 *   Title: Is palindrome
 *
 * Problem:
 *   Given a string, determine if it is a palindrome, considering only
 *   alphanumeric characters and ignoring cases.
 *
 *   Note: For the purpose of this problem, we define empty string as valid
 *   palindrome.
 *
 *   Execution: javac IsPalindrome.java && java IsPalindrome
 */
import java.util.Arrays;


public class IsPalindrome {
    public static boolean isPalindrome(String s) {
        if((s == null) || (s.length() == 0)) {
            return true;
        }

        s = String.join("",Arrays.asList(s.split("\\W+")));
        s = s.toLowerCase();

        String snew = new StringBuilder(s).reverse().toString();
        return s.equals(snew);
    }

    public static void main(String[] args) {
        assert isPalindrome("A man, a plan, a canal: Panama") == true;
        assert isPalindrome("race a car") == false;

        System.out.println("Passed all test cases");
    }
    
}

File: quickstart_guides/arrays_strings/python/is_palindrome.py
"""
Title: Is palindrome

Problem:
    Given a string, determine if it is a palindrome, considering only
    alphanumeric characters and ignoring cases.

    Note: For the purpose of this problem, we define empty string as valid
    palindrome.

Execution: python is_palindrome.py
"""
import string
import unittest


def is_palindrome(s: str) -> bool:
    s = s.lower().replace(" ", "").translate(str.maketrans("",
                                                           "",
                                                           string.punctuation))
    return True if s[::-1] == s else False


class TestIsPalindrome(unittest.TestCase):
    """Unit tests for is_palindrome."""

    def test_1(self):
        self.assertEqual(is_palindrome("A man, a plan, a canal: Panama"), True)

    def test_2(self):
        self.assertEqual(is_palindrome("race a car"), False)


if __name__ == "__main__":
    unittest.main()
